MCM - tree-traversal implementation of native MCM execution (#5180)
PennyLaneAI/pennylane: doc/introduction/measurements.rst

    If the ``defer_measurements`` transform is used in analytic mode, ``backprop`` is also a viable
    option.

> .. _tree_traversal:

Author: @mudit2812 @trbromley I'm adding this description of TT finally.

The tree-traversal algorithm
****************************

Dynamic circuit execution is akin to traversing a binary tree where each MCM
corresponds to a node and groups of gates between the MCMs correspond to edges.
The :func:`~.pennylane.dynamic_one_shot` approach picks a branch of the tree randomly
and simulates it from beginning to end.
This is wasteful in many cases; the same branch is simulated many times
when there are more shots than branches for example.
The tree-traversal algorithm does away with such redundancy while retaining the
exponential gains in memory of the one-shot approach compared with the deferred
measurement principle, among other advantages.

Briefly, it proceeds from the top of the tree and begins splitting the circuit
into two: a circuit running to the next MCM replacing the MCM by a sample

> measurement, and a circuit consisting of the rest of the circuit. A key point is that

Reviewer: I'm not sure I followed the first sentence. Though admittedly, this algorithm is always going to be hard to describe in just one paragraph!

Author: How about 
```
Briefly, it proceeds cutting an :math:`n_{MCM}` circuit into :math:`n_{MCM}+1`
circuit segments. Each segment can be executed on either the `0`- or `1`-branch,
which gives rise to a binary tree with :math:`2^{n_{MCM}}` leaves.  Terminal
measurements are obtained at the leaves, and propagated and combined back up at each
node up the tree. The tree is visited using a depth-first pattern. The tree-traversal
method improves on :func:`~.pennylane.dynamic_one_shot` by taking all samples at a
node or leaf at once. Neglecting overheads, simulating all branches requires the same
amount of computations as :func:`~. pennylane.defer_measurements`, but without the
:math:`O(2^{n_{MCM}})` memory requirement.  To save time, a copy of the state vector
is made at every branching point, or MCM, requiring at most :math:`n_{MCM}+1` state
vectors at any instant, an exponential improvement compared with :func:`~. pennylane.
defer_measurements`.  Since the counts of many nodes come out to be zero in practice,
it is often possible to ignore entire sub-trees, thereby reducing the computational
burden.
```
?

Author: @mudit2812 Do you prefer this one too?

Reviewer 2: Yeah the suggested explanation is very nice!

> into two: a circuit running to the next MCM replacing the MCM by a sample
> measurement, and a circuit consisting of the rest of the circuit. A key point is that
> all samples are collected at once, removing the need to run the circuit several times.
> The samples are converted to counts and a copy of the state vector is made at each node.
> We loop over the possible branches, collapsing the state vector and pursuing the calculation
> until the next MCM. Since many counts come out to be zero in practice, it is possible

Reviewer 2: ```suggestion
into two: a circuit running to the next mid-circuit measurement, replacing the mid-circuit measurement by a sample
measurement, and a circuit consisting of the rest of the circuit. A key point is that
all samples are collected at once, removing the need to run the circuit several times.
The samples are converted to counts and a copy of the state vector is made at each node.
We loop over the possible branches, collapsing the state vector and pursuing the calculation
until the next mid-circuit measurement. Since many counts come out to be zero in practice, it is possible
```

to ignore entire sub-trees, reducing the computational burden.

> We eventually reach the bottom of the tree having cached at most :math`n_{MCM}+1`

Reviewer: ```suggestion
We eventually reach the bottom of the tree having cached at most :math:`n_{MCM}+1`
```

state vector copies, which is an exponential improvement compared with

> :func:`~. pennylane.defer_measurements` requiring :math`O(2^{n_{MCM}})` memory.

Reviewer: ```suggestion
:func:`~. pennylane.defer_measurements` requiring :math:`O(2^{n_{MCM}})` memory.
```

We then obtain terminal measurements, propagating and combining at each node up the
tree. This algorithm is thus of the depth-first family of tree-traversal algorithms.

To summarize, this algorithm gives us the best of both worlds. In the limit of few

> shots and/or many MCMs, it is as fast as the naive shot-by-shot implementation
> because few sub-trees are explored. In the limit of many shots and/or few MCMs, it is
> equal to or faster than the deferred measurement algorithm (albeit with more
> overheads in practice) because each tree edge is visited at most once, all while
> reducing the memory requirements exponentially.
> 
> The tree-traversal algorithm is not a transform. Its usage is therefore specified
> by passing an ``mcm_method`` option to a QNode (see section
> :ref:`MCMConfig <mcm_config>`). For example,

Reviewer 2: ```suggestion
shots and/or many mid-circuit measurements, it is as fast as the naive shot-by-shot implementation
because few sub-trees are explored. In the limit of many shots and/or few mid-circuit measurements, it is
equal to or faster than the deferred measurement algorithm (albeit with more
overheads in practice) because each tree edge is visited at most once, all while
reducing the memory requirements exponentially.

The tree-traversal algorithm is not a transform. Its usage is therefore specified
by passing an ``mcm_method`` option to a QNode (see section
:ref:`"Configuring mid-circuit measurements" <mcm_config>`). For example,
```

.. code-block:: python

    @qml.qnode(dev, mcm_method="tree-traversal")
    def my_quantum_function(x, y):

>         (...)

Reviewer 2: Perhaps there should also be a note stating that this mcm_method is only available on `default.qubit` for now.

:class:`~pennylane.QNode`:

* ``mcm_method``: To set the method used for applying mid-circuit measurements. Use ``mcm_method="deferred"``
  to apply the :ref:`deferred measurements principle <deferred_measurements>`, ``mcm_method="one-shot"`` to apply
  the :ref:`one-shot transform <one_shot_transform>` or ``mcm_method="tree-traversal"`` to execute the
  :ref:`tree-traversal algorithm <tree_traversal>`.

>   When executing with finite shots, ``mcm_method="one-shot"``

Author: Yes. What do you think should be the default @trbromley ?
